---
title: "Setting up a single-cell pooled CRISPR screen"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Setting up a single-cell pooled CRISPR screen}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

This document shows how to set up a single-cell pooled CRISPR screen analysis using `ondisc`. There are five main steps:

1. Obtain the raw expression data.
2. Create an `ondisc_matrix` representing the gene expression and gRNA perturbation data.
3. Select which genes and gRNAs to use in the analysis.
4. Perform quality control on the cells.
5. Compute the cell-wise covariate matrix.

Before proceeding, we load all required packages for this example.

```{r, message=FALSE, warning=FALSE}
library(R.utils)
library(Matrix)
library(monocle)
library(readr)
library(dplyr)
library(ondisc)
```

# 1. Obtain the raw expression data

The first step is to obtain the raw expression data. We download an example single-cell CRISPR screen dataset published by Gasperini et al. in 2019. The authors searched for cis-regulatory relationships genome-wide by perturbing candidate enhancers using CRISPRi and measuring the effects of these perturbations using single-cell RNA-seq. The data consist of 32,738 genes (and long noncoding RNAs) and 3,118 gRNAs measured across 47,650 cells.

We create an example directory in which to store the data. Be sure to change the file path to a location on your machine!

```{r}
example_dir <- "/Volumes/tims_new_drive/research/sceptre_example" # change me!
if (!dir.exists(example_dir)) dir.create(path = example_dir, recursive = TRUE)
```

We additionally create a "raw data" directory and "processed data" directory as subdirectories of the example directory.

```{r}
raw_data_dir <- paste0(example_dir, "/raw_data")
processed_data_dir <- paste0(example_dir, "/processed_data")
if (!dir.exists(raw_data_dir)) dir.create(path = raw_data_dir)
if (!dir.exists(processed_data_dir)) dir.create(path = processed_data_dir)
```

Next, we download the data.

```{r, eval=FALSE, echo=TRUE}
# set remote URL
remote <- "https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE120861&format=file&file="
# set file names
file_names <- c("GSE120861_pilot_highmoi_screen.cds.rds",
                "GSE120861_pilot_highmoi_screen.cells.txt",
                "GSE120861_pilot_highmoi_screen.exprs.mtx",
                "GSE120861_pilot_highmoi_screen.genes.txt",
                "GSE120861_pilot_highmoi_screen.phenoData.txt",
                "GSE120861_gene_gRNAgroup_pair_table.pilot.txt")
# allow downloads of up to 10 minutes
options(timeout = 10*60)
# download
for (file_name in file_names) {
  dest <- paste0(raw_data_dir, "/", file_name)
  dest_gz <- paste0(dest, ".gz")
  if (!file.exists(dest)) {
    cat(paste0("Downloading ", file_name, "\n"))
    source <- paste0(remote, file_name, ".gz")
    download.file(url = source, destfile = dest_gz)
    if (file.exists(dest_gz)) gunzip(dest_gz)
  }
}
```

# 2. Convert the gene and gRNA data into an `ondisc_matrix`

Next, we convert the gene expression and gRNA indicator data into an `ondisc_matrix` object. First, we call `create_ondisc_matrix_from_mtx` on the gene expression data. This function takes about 15-20 minutes to run on a Macbook Pro laptop. (In general, it takes about 8 minutes per gigabyte to run; we are working to make this even faster.)

```{r, eval=FALSE, echo=TRUE}
features_fp <- paste0(raw_data_dir, "/GSE120861_pilot_highmoi_screen.genes.txt")
cell_fp <- paste0(raw_data_dir, "/GSE120861_pilot_highmoi_screen.cells.txt")
mtx_fp <- paste0(raw_data_dir, "/GSE120861_pilot_highmoi_screen.exprs.mtx")
expressions <- create_ondisc_matrix_from_mtx(mtx_fp = mtx_fp,
                                             barcodes_fp = cell_fp,
                                             features_fp = features_fp,
                                             n_lines_per_chunk = 35000000,
                                             on_disk_dir = processed_data_dir,
                                             return_metadata_ondisc_matrix = TRUE)
expressions
saveRDS(expressions, paste0(raw_data_dir, "/expressions.rds"))
```

```{r, echo=FALSE}
expressions <- readRDS(paste0(raw_data_dir, "/expressions.rds"))
```

The variable `expressions` is a variable of class `metadata_ondisc_matrix`, containing the fields `ondisc_matrix`, `cell_covariates`, and `feature_covariates`.

Next, we put the gRNA data into an `ondisc_matrix`. Unfortunately, the gRNA data are not in a standard format: they are stored a [Monocle](https://cole-trapnell-lab.github.io/monocle3/) R object. Thus, we load the Monocole object into R, extract the gRNA data, and write the data to a logical (i.e., binary) .mtx file called `perturbations.mtx`. Additionally, we write the gRNA IDs to a .tsv file called `gRNAs.tsv`.

```{r,eval=FALSE,echo=TRUE}
x <- readRDS(paste0(raw_data_dir, "/", "GSE120861_pilot_highmoi_screen.cds.rds"))
cell_covs <- pData(x)
gRNA_cols <- 15:ncol(cell_covs)
gRNA_names <- colnames(cell_covs[,gRNA_cols])
gRNA_indics <- as.matrix(cell_covs[,gRNA_cols])
colnames(gRNA_indics) <- row.names(gRNA_indics) <- NULL
sparse_gRNA_indics <- t(Matrix(gRNA_indics, sparse = TRUE))
writeMM(obj = sparse_gRNA_indics, file = paste0(raw_data_dir, "/perturbations.mtx"))
gRNA_features <- data.frame(gRNA_id = gRNA_names)
write_tsv(x = gRNA_features, file = paste0(raw_data_dir, "/gRNAs.tsv"), col_names = FALSE)
rm(x, cell_covs, gRNA_cols, gRNA_names, gRNA_indics, sparse_gRNA_indics, gRNA_features)
```

Now, we can convert the perturbation data into an `ondisc_matrix` by calling `create_ondisc_matrix_from_mtx` on `perturbations.mtx` and its associated metadata files. The function takes a few seconds to run.

```{r,eval=FALSE,echo=TRUE}
features_fp <- paste0(raw_data_dir, "/gRNAs.tsv")
cell_fp <- paste0(raw_data_dir, "/GSE120861_pilot_highmoi_screen.cells.txt")
mtx_fp <- paste0(raw_data_dir, "/perturbations.mtx")
perturbations <- create_ondisc_matrix_from_mtx(mtx_fp = mtx_fp,
                                               barcodes_fp = cell_fp,
                                               features_fp = features_fp,
                                               on_disk_dir = processed_data_dir,
                                               return_metadata_ondisc_matrix = TRUE,
                                               progress = FALSE)
perturbations
saveRDS(object = perturbations, file = paste0(raw_data_dir, "/perturbations.rds"))
```

```{r,echo=FALSE}
perturbations <- readRDS(file = paste0(raw_data_dir, "/perturbations.rds"))
```

**NOTE**: For a given cell and gRNA, the example data do not report the actual number of gRNA transcripts (or barcodes) detected in the cell. Rather, the data report a thresholded indicator variable: 1 if the number of detected gRNA UMIs exceeds four, and 0 otherwise. If you are working with raw gRNA UMI counts, you will need to threshold the gRNA counts before you initialize the perturbation `ondisc_matrix`. We will add functionality to automatically threshold the gRNA UMI counts in a future update.

Also note that we say that a cell is *perturbed* by a gRNA if the corresponding thresholded indicator variable is 1, and *unperturbed* otherwise.

# 3. Select genes and gRNAs to use in the analysis

We select the genes, gRNAs, and gene-gRNA pairs to use in the analysis. There are no hard-and-fast guidelines here, but as a starting point, we exclude genes that are expressed in fewer than 1,000 cells and gRNAs that are expressed in fewer than 100 cells. We ultimately keep 10,870 of the 32,738 genes and 1,560 of the 1,561 gRNAs. 

```{r}
# subset expressions matrix
genes_to_keep <- expressions@feature_covariates$n_nonzero > 1000
expressions_sub <- expressions[genes_to_keep,]

# subset perturbations matrix
gRNAs_to_keep <- perturbations@feature_covariates$n_nonzero > 100
perturbations_sub <- perturbations[gRNAs_to_keep,]
```

Next, we select the gene-gRNA pairs to analyze. The file `GSE120861_gene_gRNAgroup_pair_table.pilot.txt` contains a list of all the gene-gRNA pairs that were analyzed in the original study. The gRNAs are grouped into several categories, including non-targeting controls, transcription start sites, and DNase I hypersensitive sites (i.e., the candidate enhancer locations). First, we intersect the gene-gRNA pairs with the highly-expressed genes and gRNAs that we retained in the previous step. Next, we randomly sample 500 of the non-targeting control pairs for inclusion in the analysis. (This is for demonstration purposes only; if our goal were to assess calibration, we would want to sample the non-targeting controls in a more careful way.)

```{r}
# load the gene-gRNA pairs used in the original analysis
original_pairs <- read_tsv(file = paste0(raw_data_dir,
                                         "/GSE120861_gene_gRNAgroup_pair_table.pilot.txt"),
                           col_types = c("___cc___c____"), col_names = TRUE) %>% 
  rename("gRNA_id" = "gRNAgroup",
         "gene_id" = "ENSG.targetgene",
         "gRNA_type" = "general_group")

# get the IDs of the highly expressed genes and gRNAs
genes_in_analysis <- get_feature_ids(expressions)
gRNAs_in_analysis <- get_feature_ids(perturbations)

# filter and sample
set.seed(100)
gene_gRNA_pairs <- filter(.data = original_pairs,
                          gRNA_id %in% gRNAs_in_analysis && 
                            gene_id %in% genes_in_analysis) %>%
  filter(gRNA_type %in% "NTC") %>% slice_sample(n = 500)
```

Finally, we save the gene-gRNA pairs that we have selected for analysis in the "processed data" directory.

```{r}
saveRDS(object = gene_gRNA_pairs, file = paste0(processed_data_dir, "/gene_gRNA_pairs.rds"))
```

# 4. Perform quality control on the cells

The next step is to perform quality control on the cells. First, we create a [multimodal_ondisc_matrix](https://timothy-barry.github.io/ondisc/articles/tutorial_other_classes.html), combining the `expressions` and `perturbations` modalities into a single object. 

```{r}
crispr_experiment <- multimodal_ondisc_matrix(list(expressions = expressions_sub,
                                                   perturbations = perturbations_sub))
```

Using the raw, unfiltered global-cell covariate matrix, we exclude cells that received zero perturbations.

```{r}
cells_to_keep <- crispr_experiment@global_cell_covariates$perturbations_n_nonzero > 0
crispr_experiment_sub <- crispr_experiment[,cells_to_keep]
```

# 5. Compute the cell-wise covariate matrix

Finally, we compute the cell-wise covariate matrix. The following cell-specific covariates are reasonable choices to include in the analysis:

1. sequencing batch
2. percentage of UMIs that map to mitochondrial genes
3. log-transformed mRNA UMI count
4. log-transformed number of genes expressed in cell
5. log-transformed number of perturbations detected in cell, or log-transfored gRNA UMI count.

In general, it probably is best to include as many of the above covariates in the analysis as possible. In our current example we can compute covariates 3-5:

```{r}
cell_covariate_matrix <- summarize(crispr_experiment_sub@global_cell_covariates,
                                lg_umi_count = log(expressions_n_umis),
                                lg_n_genes_expressed = log(expressions_n_nonzero),
                                lg_n_perturbations = log(perturbations_n_nonzero))
```

We save the global cell-covariate matrix and the appropriately-subsetted gene and perturbation `ondisc_matrices` to the "processed data directory."

```{r}
saveRDS(object = cell_covariate_matrix,
        file = paste0(processed_data_dir, "/cell_covariate_matrix.rds"))
saveRDS(object = crispr_experiment_sub@modalities$expressions,
        file = paste0(processed_data_dir, "/expressions.rds"))
saveRDS(object = crispr_experiment_sub@modalities$perturbations,
        file = paste0(processed_data_dir, "/perturbations.rds"))
```

The "processed data directory" should contain the following files:

1. The expressions.rds `ondisc_matrix` plus its backing .h5 file,
2. The perturbations.rds `ondisc_matsrix` plus its backing .h5 file,
3. The cell_covariate_matrix.rds file, and
4. The gene_gRNA_pairs.rds file.

Now, we are ready to run our single-cell CRISPR screen analysis!
