---
title: "Setting up a single-cell pooled CRISPR screen analysis"
---

```{r, include = FALSE, eval=FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

This document shows how to set up a pooled CRISPR screen analysis using `ondisc`. There are five main steps:

1. Obtain the raw expression data.
2. Create an `ondisc_matrix` representing the gene expressions and gRNA indicators.
3. Select which genes and gRNAs to use in the analysis.
4. Perform quality control on the cells.
5. Compute the cell-wise covariate matrix.

Before proceeding, we load all required packages for this example.

```{r, echo=FALSE,warning=FALSE,message=FALSE,eval=FALSE}
library(R.utils)
library(Matrix)
library(monocle)
library(readr)
library(ondisc)
```

# 1. Obtain the raw expression data

The first step is to obtain the raw expression data in the form of UMI count matrices. We download an example single-cell CRISPR screen dataset from the GEO database, published by Gasperini et al. in 2019. The example consists of 32,738 genes (and long noncoding RNAs), 47,650 cells, and 3,118 gRNAs.

We set a directory in which to store the data. Be sure to change the file path to a location on your machine!

```{r,eval=FALSE}
example_dir <- "/Volumes/tims_new_drive/research/sceptre_example" # change me!
if (!dir.exists(example_dir)) dir.create(path = example_dir, recursive = TRUE)
```

Next, we download the data.

```{r, eval=FALSE,echo=FALSE}
raw_data_dir <- paste0(example_dir, "/raw_data")
processed_data_dir <- paste0(example_dir, "/processed_data")
if (!dir.exists(raw_data_dir)) dir.create(path = raw_data_dir)
if (!dir.exists(processed_data_dir)) dir.create(path = processed_data_dir)
# set remote URL
remote <- "https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE120861&format=file&file="
# set file names
file_names <- c("GSE120861_pilot_highmoi_screen.cds.rds",
                "GSE120861_pilot_highmoi_screen.cells.txt",
                "GSE120861_pilot_highmoi_screen.exprs.mtx",
                "GSE120861_pilot_highmoi_screen.genes.txt",
                "GSE120861_pilot_highmoi_screen.phenoData.txt")
# allow downloads of up to 10 minutes
options(timeout = 10*60)
# download
for (file_name in file_names) {
  dest <- paste0(raw_data_dir, "/", file_name)
  dest_gz <- paste0(dest, ".gz")
  if (!file.exists(dest)) {
    cat(paste0("Downloading ", file_name, "\n"))
    source <- paste0(remote, file_name, ".gz")
    download.file(url = source, destfile = dest_gz)
    if (file.exists(dest_gz)) gunzip(dest_gz)
  }
}
```

# 2. Put the gene and gRNA data into an `ondisc_matrix`

Now that we have downloaded the data, we can put the gene expression and gRNA indicator data into `ondisc_matrix` objects. First, we call `create_ondisc_matrix_from_mtx` on the gene expression data (which are roughly 2GB in size). This function takes about 2 minutes to run on a Macbook Pro laptop.

```{r,eval=FALSE}
features_fp <- paste0(raw_data_dir, "/GSE120861_pilot_highmoi_screen.genes.txt")
cell_fp <- paste0(raw_data_dir, "/GSE120861_pilot_highmoi_screen.cells.txt")
mtx_fp <- paste0(raw_data_dir, "/GSE120861_pilot_highmoi_screen.exprs.mtx")
expressions <- create_ondisc_matrix_from_mtx(mtx_fp = mtx_fp,
                                             barcodes_fp = cell_fp,
                                             features_fp = feature_fps,
                                             n_lines_per_chunk = 90000000,
                                             on_disc_dir = processed_data_dir,
                                             return_metadata_ondisc_matrix = TRUE)
expressions
```

The variable `expressions` is a variable of class `metadata_ondisc_matrix`. It contains three fields:

1. an `ondisc_matrix` representing the expression data (access through `expressions@ondisc_matrix`)
2. a data frame representing the cell-specific covariates (access through `expressions@cell_covariates`)
3. a data frame representing the gene-specific covariates (access through `expressions@feature_covariates`)

We save `expressions` as an RDS file so that we can access it later.

```{r,eval=FALSE}
saveRDS(expressions, paste0(processed_data_dir, "/expressions.rds"))
```

Next, we put the gRNA data into an `ondisc_matrix`. Unfortunately, the gRNA data are not in a standard format; instead, they come in the form of a [Monocle](https://cole-trapnell-lab.github.io/monocle3/) R object. Thus, we load the Monocole object, extract the gRNA data, and write the data to a logical (i.e., binary) .mtx file called `perturbations.mtx`. Additionally, we write the gRNA IDs to a .tsv file called `gRNAs.tsv`.

```{r,eval=FALSE}
x <- readRDS(paste0(raw_data_dir, "/", "GSE120861_pilot_highmoi_screen.cds.rds"))
cell_covs <- pData(x)
gRNA_cols <- 15:ncol(cell_covs)
gRNA_names <- colnames(cell_covs[,gRNA_cols])
gRNA_indics <- as.matrix(cell_covs[,gRNA_cols])
colnames(gRNA_indics) <- row.names(gRNA_indics) <- NULL
sparse_gRNA_indics <- t(Matrix(gRNA_indics, sparse = TRUE))
writeMM(obj = sparse_gRNA_indics, file = paste0(raw_data_dir, "/perturbations.mtx"))
gRNA_features <- tibble(gRNA_id = gRNA_names)
write_tsv(x = gRNA_features, file = paste0(raw_data_dir, "/gRNAs.tsv"), col_names = FALSE)
rm(x, cell_covs, gRNA_cols, gRNA_names, gRNA_indics, sparse_gRNA_indics, gRNA_features)
```

Now we can put the perturbation data into an `ondisc_matrix` by calling `create_ondisc_matrix_from_mtx`. The function takes a few seconds to run.

```{r,eval=FALSE}
features_fp <- paste0(raw_data_dir, "/gRNAs.tsv")
cell_fp <- paste0(raw_data_dir, "/GSE120861_pilot_highmoi_screen.cells.txt")
mtx_fp <- paste0(raw_data_dir, "/perturbations.mtx")
perturbations <- create_ondisc_matrix_from_mtx(mtx_fp = mtx_fp,
                                               barcodes_fp = cell_fp,
                                               features_fp = features_fp,
                                               on_disc_dir = processed_data_dir,
                                               return_metadata_ondisc_matrix = TRUE)
perturbations
saveRDS(perturbations, paste0(processed_data_dir, "/perturbations.rds"))
```

# 3. Select genes and gRNAs to use in the analysis

```{r,eval=FALSE}

```
