<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Setting up a single-cell pooled CRISPR screen analysis • ondisc</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/lumen/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../../bootstrap-toc.css">
<script src="../../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../../pkgdown.css" rel="stylesheet">
<script src="../../pkgdown.js"></script><meta property="og:title" content="Setting up a single-cell pooled CRISPR screen analysis">
<meta property="og:description" content="ondisc">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../../index.html">ondisc</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.9.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../../reference/index.html">Function reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/Timothy-Barry/ondisc/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="setting_up_a_crispr_screen_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Setting up a single-cell pooled CRISPR screen analysis</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/Timothy-Barry/ondisc/blob/master/vignettes/articles/setting_up_a_crispr_screen.Rmd"><code>vignettes/articles/setting_up_a_crispr_screen.Rmd</code></a></small>
      <div class="hidden name"><code>setting_up_a_crispr_screen.Rmd</code></div>

    </div>

    
    
<p>This document shows how to set up a pooled CRISPR screen analysis using <code>ondisc</code>. There are five main steps:</p>
<ol style="list-style-type: decimal">
<li>Obtain the raw expression data.</li>
<li>Create an <code>ondisc_matrix</code> representing the gene expressions and gRNA indicators.</li>
<li>Select which genes and gRNAs to use in the analysis.</li>
<li>Perform quality control on the cells.</li>
<li>Compute the cell-wise covariate matrix.</li>
</ol>
<p>Before proceeding, we load all required packages for this example.</p>
<div id="obtain-the-raw-expression-data" class="section level1">
<h1 class="hasAnchor">
<a href="#obtain-the-raw-expression-data" class="anchor"></a>1. Obtain the raw expression data</h1>
<p>The first step is to obtain the raw expression data in the form of UMI count matrices. We download an example single-cell CRISPR screen dataset from the GEO database, published by Gasperini et al. in 2019. The example consists of 32,738 genes (and long noncoding RNAs), 47,650 cells, and 3,118 gRNAs.</p>
<p>We set a directory in which to store the data. Be sure to change the file path to a location on your machine!</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">example_dir</span> <span class="op">&lt;-</span> <span class="st">"/Volumes/tims_new_drive/research/sceptre_example"</span> <span class="co"># change me!</span>
<span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files2.html">dir.exists</a></span><span class="op">(</span><span class="va">example_dir</span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/files2.html">dir.create</a></span><span class="op">(</span>path <span class="op">=</span> <span class="va">example_dir</span>, recursive <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p>Next, we download the data.</p>
</div>
<div id="put-the-gene-and-grna-data-into-an-ondisc_matrix" class="section level1">
<h1 class="hasAnchor">
<a href="#put-the-gene-and-grna-data-into-an-ondisc_matrix" class="anchor"></a>2. Put the gene and gRNA data into an <code>ondisc_matrix</code>
</h1>
<p>Now that we have downloaded the data, we can put the gene expression and gRNA indicator data into <code>ondisc_matrix</code> objects. First, we call <code>create_ondisc_matrix_from_mtx</code> on the gene expression data (which are roughly 2GB in size). This function takes about 2 minutes to run on a Macbook Pro laptop.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">features_fp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">raw_data_dir</span>, <span class="st">"/GSE120861_pilot_highmoi_screen.genes.txt"</span><span class="op">)</span>
<span class="va">cell_fp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">raw_data_dir</span>, <span class="st">"/GSE120861_pilot_highmoi_screen.cells.txt"</span><span class="op">)</span>
<span class="va">mtx_fp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">raw_data_dir</span>, <span class="st">"/GSE120861_pilot_highmoi_screen.exprs.mtx"</span><span class="op">)</span>
<span class="va">expressions</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/create_ondisc_matrix_from_mtx.html">create_ondisc_matrix_from_mtx</a></span><span class="op">(</span>mtx_fp <span class="op">=</span> <span class="va">mtx_fp</span>,
                                             barcodes_fp <span class="op">=</span> <span class="va">cell_fp</span>,
                                             features_fp <span class="op">=</span> <span class="va">feature_fps</span>,
                                             n_lines_per_chunk <span class="op">=</span> <span class="fl">90000000</span>,
                                             on_disc_dir <span class="op">=</span> <span class="va">processed_data_dir</span>,
                                             return_metadata_ondisc_matrix <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="va">expressions</span></code></pre></div>
<p>The variable <code>expressions</code> is a variable of class <code>metadata_ondisc_matrix</code>. It contains three fields:</p>
<ol style="list-style-type: decimal">
<li>an <code>ondisc_matrix</code> representing the expression data (access through <code>expressions@ondisc_matrix</code>)</li>
<li>a data frame representing the cell-specific covariates (access through <code>expressions@cell_covariates</code>)</li>
<li>a data frame representing the gene-specific covariates (access through <code>expressions@feature_covariates</code>)</li>
</ol>
<p>We save <code>expressions</code> as an RDS file so that we can access it later.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">saveRDS</a></span><span class="op">(</span><span class="va">expressions</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">processed_data_dir</span>, <span class="st">"/expressions.rds"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Next, we put the gRNA data into an <code>ondisc_matrix</code>. Unfortunately, the gRNA data are not in a standard format; instead, they come in the form of a <a href="https://cole-trapnell-lab.github.io/monocle3/">Monocle</a> R object. Thus, we load the Monocole object, extract the gRNA data, and write the data to a logical (i.e., binary) .mtx file called <code>perturbations.mtx</code>. Additionally, we write the gRNA IDs to a .tsv file called <code>gRNAs.tsv</code>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">raw_data_dir</span>, <span class="st">"/"</span>, <span class="st">"GSE120861_pilot_highmoi_screen.cds.rds"</span><span class="op">)</span><span class="op">)</span>
<span class="va">cell_covs</span> <span class="op">&lt;-</span> <span class="fu">pData</span><span class="op">(</span><span class="va">x</span><span class="op">)</span>
<span class="va">gRNA_cols</span> <span class="op">&lt;-</span> <span class="fl">15</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">cell_covs</span><span class="op">)</span>
<span class="va">gRNA_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">cell_covs</span><span class="op">[</span>,<span class="va">gRNA_cols</span><span class="op">]</span><span class="op">)</span>
<span class="va">gRNA_indics</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">cell_covs</span><span class="op">[</span>,<span class="va">gRNA_cols</span><span class="op">]</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">gRNA_indics</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/row.names.html">row.names</a></span><span class="op">(</span><span class="va">gRNA_indics</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="cn">NULL</span>
<span class="va">sparse_gRNA_indics</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="fu">Matrix</span><span class="op">(</span><span class="va">gRNA_indics</span>, sparse <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
<span class="fu">writeMM</span><span class="op">(</span>obj <span class="op">=</span> <span class="va">sparse_gRNA_indics</span>, file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">raw_data_dir</span>, <span class="st">"/perturbations.mtx"</span><span class="op">)</span><span class="op">)</span>
<span class="va">gRNA_features</span> <span class="op">&lt;-</span> <span class="fu">tibble</span><span class="op">(</span>gRNA_id <span class="op">=</span> <span class="va">gRNA_names</span><span class="op">)</span>
<span class="fu">write_tsv</span><span class="op">(</span>x <span class="op">=</span> <span class="va">gRNA_features</span>, file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">raw_data_dir</span>, <span class="st">"/gRNAs.tsv"</span><span class="op">)</span>, col_names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/rm.html">rm</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">cell_covs</span>, <span class="va">gRNA_cols</span>, <span class="va">gRNA_names</span>, <span class="va">gRNA_indics</span>, <span class="va">sparse_gRNA_indics</span>, <span class="va">gRNA_features</span><span class="op">)</span></code></pre></div>
<p>Now we can put the perturbation data into an <code>ondisc_matrix</code> by calling <code>create_ondisc_matrix_from_mtx</code>. The function takes a few seconds to run.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">features_fp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">raw_data_dir</span>, <span class="st">"/gRNAs.tsv"</span><span class="op">)</span>
<span class="va">cell_fp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">raw_data_dir</span>, <span class="st">"/GSE120861_pilot_highmoi_screen.cells.txt"</span><span class="op">)</span>
<span class="va">mtx_fp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">raw_data_dir</span>, <span class="st">"/perturbations.mtx"</span><span class="op">)</span>
<span class="va">perturbations</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/create_ondisc_matrix_from_mtx.html">create_ondisc_matrix_from_mtx</a></span><span class="op">(</span>mtx_fp <span class="op">=</span> <span class="va">mtx_fp</span>,
                                               barcodes_fp <span class="op">=</span> <span class="va">cell_fp</span>,
                                               features_fp <span class="op">=</span> <span class="va">features_fp</span>,
                                               on_disc_dir <span class="op">=</span> <span class="va">processed_data_dir</span>,
                                               return_metadata_ondisc_matrix <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="va">perturbations</span>
<span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">saveRDS</a></span><span class="op">(</span><span class="va">perturbations</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">processed_data_dir</span>, <span class="st">"/perturbations.rds"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div id="select-genes-and-grnas-to-use-in-the-analysis" class="section level1">
<h1 class="hasAnchor">
<a href="#select-genes-and-grnas-to-use-in-the-analysis" class="anchor"></a>3. Select genes and gRNAs to use in the analysis</h1>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://timothy-barry.github.io/">Tim Barry</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
