<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Setting up a single-cell pooled CRISPR screen • ondisc</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/lumen/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Setting up a single-cell pooled CRISPR screen">
<meta property="og:description" content="ondisc">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ondisc</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Function reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/tutorial_odm_class.html">Tutorial 1</a>
    </li>
    <li>
      <a href="../articles/tutorial_other_classes.html">Tutorial 2</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Ondisc in action
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/setting_up_a_crispr_screen.html">Setting up a single-cell CRISPR screen</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="setting_up_a_crispr_screen_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Setting up a single-cell pooled CRISPR screen</h1>
            
      
      
      <div class="hidden name"><code>setting_up_a_crispr_screen.Rmd</code></div>

    </div>

    
    
<p>This document shows how to set up a single-cell pooled CRISPR screen analysis using <code>ondisc</code>. There are five main steps:</p>
<ol style="list-style-type: decimal">
<li>Obtain the raw expression data.</li>
<li>Create an <code>ondisc_matrix</code> representing the gene expression and gRNA perturbation data.</li>
<li>Select which genes and gRNAs to use in the analysis.</li>
<li>Perform quality control on the cells.</li>
<li>Compute the cell-wise covariate matrix.</li>
</ol>
<p>Before proceeding, we load all required packages for this example.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/HenrikBengtsson/R.utils">R.utils</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://Matrix.R-forge.R-project.org/">Matrix</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">monocle</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://readr.tidyverse.org">readr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://timothy-barry.github.io/ondisc/">ondisc</a></span><span class="op">)</span></code></pre></div>
<div id="obtain-the-raw-expression-data" class="section level1">
<h1 class="hasAnchor">
<a href="#obtain-the-raw-expression-data" class="anchor"></a>1. Obtain the raw expression data</h1>
<p>The first step is to obtain the raw expression data. We download an example single-cell CRISPR screen dataset published by Gasperini et al. in 2019. The authors searched for cis-regulatory relationships genome-wide by perturbing candidate enhancers using CRISPRi and measuring the effects of these perturbations using single-cell RNA-seq. The data consist of 32,738 genes (and long noncoding RNAs) and 3,118 gRNAs measured across 47,650 cells.</p>
<p>We create an example directory in which to store the data. Be sure to change the file path to a location on your machine!</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">example_dir</span> <span class="op">&lt;-</span> <span class="st">"/Volumes/tims_new_drive/research/sceptre_example"</span> <span class="co"># change me!</span>
<span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files2.html">dir.exists</a></span><span class="op">(</span><span class="va">example_dir</span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/files2.html">dir.create</a></span><span class="op">(</span>path <span class="op">=</span> <span class="va">example_dir</span>, recursive <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p>We additionally create a “raw data” directory and “processed data” directory as subdirectories of the example directory.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">raw_data_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">example_dir</span>, <span class="st">"/raw_data"</span><span class="op">)</span>
<span class="va">processed_data_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">example_dir</span>, <span class="st">"/processed_data"</span><span class="op">)</span>
<span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files2.html">dir.exists</a></span><span class="op">(</span><span class="va">raw_data_dir</span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/files2.html">dir.create</a></span><span class="op">(</span>path <span class="op">=</span> <span class="va">raw_data_dir</span><span class="op">)</span>
<span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files2.html">dir.exists</a></span><span class="op">(</span><span class="va">processed_data_dir</span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/files2.html">dir.create</a></span><span class="op">(</span>path <span class="op">=</span> <span class="va">processed_data_dir</span><span class="op">)</span></code></pre></div>
<p>Next, we download the data.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># set remote URL</span>
<span class="va">remote</span> <span class="op">&lt;-</span> <span class="st">"https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE120861&amp;format=file&amp;file="</span>
<span class="co"># set file names</span>
<span class="va">file_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"GSE120861_pilot_highmoi_screen.cds.rds"</span>,
                <span class="st">"GSE120861_pilot_highmoi_screen.cells.txt"</span>,
                <span class="st">"GSE120861_pilot_highmoi_screen.exprs.mtx"</span>,
                <span class="st">"GSE120861_pilot_highmoi_screen.genes.txt"</span>,
                <span class="st">"GSE120861_pilot_highmoi_screen.phenoData.txt"</span>,
                <span class="st">"GSE120861_gene_gRNAgroup_pair_table.pilot.txt"</span><span class="op">)</span>
<span class="co"># allow downloads of up to 10 minutes</span>
<span class="fu"><a href="https://rdrr.io/r/base/options.html">options</a></span><span class="op">(</span>timeout <span class="op">=</span> <span class="fl">10</span><span class="op">*</span><span class="fl">60</span><span class="op">)</span>
<span class="co"># download</span>
<span class="kw">for</span> <span class="op">(</span><span class="va">file_name</span> <span class="kw">in</span> <span class="va">file_names</span><span class="op">)</span> <span class="op">{</span>
  <span class="va">dest</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">raw_data_dir</span>, <span class="st">"/"</span>, <span class="va">file_name</span><span class="op">)</span>
  <span class="va">dest_gz</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">dest</span>, <span class="st">".gz"</span><span class="op">)</span>
  <span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files.html">file.exists</a></span><span class="op">(</span><span class="va">dest</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
    <span class="fu"><a href="https://rdrr.io/pkg/R.utils/man/Non-documented_objects.html">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"Downloading "</span>, <span class="va">file_name</span>, <span class="st">"\n"</span><span class="op">)</span><span class="op">)</span>
    <span class="va">source</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">remote</span>, <span class="va">file_name</span>, <span class="st">".gz"</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/download.file.html">download.file</a></span><span class="op">(</span>url <span class="op">=</span> <span class="va">source</span>, destfile <span class="op">=</span> <span class="va">dest_gz</span><span class="op">)</span>
    <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/files.html">file.exists</a></span><span class="op">(</span><span class="va">dest_gz</span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/pkg/R.utils/man/compressFile.html">gunzip</a></span><span class="op">(</span><span class="va">dest_gz</span><span class="op">)</span>
  <span class="op">}</span>
<span class="op">}</span></code></pre></div>
</div>
<div id="convert-the-gene-and-grna-data-into-an-ondisc_matrix" class="section level1">
<h1 class="hasAnchor">
<a href="#convert-the-gene-and-grna-data-into-an-ondisc_matrix" class="anchor"></a>2. Convert the gene and gRNA data into an <code>ondisc_matrix</code>
</h1>
<p>Next, we convert the gene expression and gRNA indicator data into an <code>ondisc_matrix</code> object. First, we call <code>create_ondisc_matrix_from_mtx</code> on the gene expression data. This function takes about 15-20 minutes to run on a Macbook Pro laptop. (In general, it takes about 8 minutes per gigabyte to run; we are working to make this even faster.)</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">features_fp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">raw_data_dir</span>, <span class="st">"/GSE120861_pilot_highmoi_screen.genes.txt"</span><span class="op">)</span>
<span class="va">cell_fp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">raw_data_dir</span>, <span class="st">"/GSE120861_pilot_highmoi_screen.cells.txt"</span><span class="op">)</span>
<span class="va">mtx_fp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">raw_data_dir</span>, <span class="st">"/GSE120861_pilot_highmoi_screen.exprs.mtx"</span><span class="op">)</span>
<span class="va">expressions</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_ondisc_matrix_from_mtx.html">create_ondisc_matrix_from_mtx</a></span><span class="op">(</span>mtx_fp <span class="op">=</span> <span class="va">mtx_fp</span>,
                                             barcodes_fp <span class="op">=</span> <span class="va">cell_fp</span>,
                                             features_fp <span class="op">=</span> <span class="va">features_fp</span>,
                                             n_lines_per_chunk <span class="op">=</span> <span class="fl">35000000</span>,
                                             on_disk_dir <span class="op">=</span> <span class="va">processed_data_dir</span>,
                                             return_metadata_ondisc_matrix <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="va">expressions</span>
<span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">saveRDS</a></span><span class="op">(</span><span class="va">expressions</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">raw_data_dir</span>, <span class="st">"/expressions.rds"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>The variable <code>expressions</code> is a variable of class <code>metadata_ondisc_matrix</code>, containing the fields <code>ondisc_matrix</code>, <code>cell_covariates</code>, and <code>feature_covariates</code>.</p>
<p>Next, we put the gRNA data into an <code>ondisc_matrix</code>. Unfortunately, the gRNA data are not in a standard format: they are stored a <a href="https://cole-trapnell-lab.github.io/monocle3/">Monocle</a> R object. Thus, we load the Monocole object into R, extract the gRNA data, and write the data to a logical (i.e., binary) .mtx file called <code>perturbations.mtx</code>. Additionally, we write the gRNA IDs to a .tsv file called <code>gRNAs.tsv</code>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">raw_data_dir</span>, <span class="st">"/"</span>, <span class="st">"GSE120861_pilot_highmoi_screen.cds.rds"</span><span class="op">)</span><span class="op">)</span>
<span class="va">cell_covs</span> <span class="op">&lt;-</span> <span class="fu">pData</span><span class="op">(</span><span class="va">x</span><span class="op">)</span>
<span class="va">gRNA_cols</span> <span class="op">&lt;-</span> <span class="fl">15</span><span class="op">:</span><span class="fu"><a href="../reference/dim.html">ncol</a></span><span class="op">(</span><span class="va">cell_covs</span><span class="op">)</span>
<span class="va">gRNA_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">cell_covs</span><span class="op">[</span>,<span class="va">gRNA_cols</span><span class="op">]</span><span class="op">)</span>
<span class="va">gRNA_indics</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">cell_covs</span><span class="op">[</span>,<span class="va">gRNA_cols</span><span class="op">]</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">gRNA_indics</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/row.names.html">row.names</a></span><span class="op">(</span><span class="va">gRNA_indics</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="cn">NULL</span>
<span class="va">sparse_gRNA_indics</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/Matrix.html">Matrix</a></span><span class="op">(</span><span class="va">gRNA_indics</span>, sparse <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/externalFormats.html">writeMM</a></span><span class="op">(</span>obj <span class="op">=</span> <span class="va">sparse_gRNA_indics</span>, file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">raw_data_dir</span>, <span class="st">"/perturbations.mtx"</span><span class="op">)</span><span class="op">)</span>
<span class="va">gRNA_features</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>gRNA_id <span class="op">=</span> <span class="va">gRNA_names</span><span class="op">)</span>
<span class="fu"><a href="https://readr.tidyverse.org/reference/write_delim.html">write_tsv</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">gRNA_features</span>, file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">raw_data_dir</span>, <span class="st">"/gRNAs.tsv"</span><span class="op">)</span>, col_names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/rm.html">rm</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">cell_covs</span>, <span class="va">gRNA_cols</span>, <span class="va">gRNA_names</span>, <span class="va">gRNA_indics</span>, <span class="va">sparse_gRNA_indics</span>, <span class="va">gRNA_features</span><span class="op">)</span></code></pre></div>
<p>Now, we can convert the perturbation data into an <code>ondisc_matrix</code> by calling <code>create_ondisc_matrix_from_mtx</code> on <code>perturbations.mtx</code> and its associated metadata files. The function takes a few seconds to run.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">features_fp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">raw_data_dir</span>, <span class="st">"/gRNAs.tsv"</span><span class="op">)</span>
<span class="va">cell_fp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">raw_data_dir</span>, <span class="st">"/GSE120861_pilot_highmoi_screen.cells.txt"</span><span class="op">)</span>
<span class="va">mtx_fp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">raw_data_dir</span>, <span class="st">"/perturbations.mtx"</span><span class="op">)</span>
<span class="va">perturbations</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_ondisc_matrix_from_mtx.html">create_ondisc_matrix_from_mtx</a></span><span class="op">(</span>mtx_fp <span class="op">=</span> <span class="va">mtx_fp</span>,
                                               barcodes_fp <span class="op">=</span> <span class="va">cell_fp</span>,
                                               features_fp <span class="op">=</span> <span class="va">features_fp</span>,
                                               on_disk_dir <span class="op">=</span> <span class="va">processed_data_dir</span>,
                                               return_metadata_ondisc_matrix <span class="op">=</span> <span class="cn">TRUE</span>,
                                               progress <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="va">perturbations</span>
<span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">saveRDS</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">perturbations</span>, file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">raw_data_dir</span>, <span class="st">"/perturbations.rds"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><strong>NOTE</strong>: For a given cell and gRNA, the example data do not report the actual number of gRNA transcripts (or barcodes) detected in the cell. Rather, the data report a thresholded indicator variable: 1 if the number of detected gRNA UMIs exceeds four, and 0 otherwise. If you are working with raw gRNA UMI counts, you will need to threshold the gRNA counts before you initialize the perturbation <code>ondisc_matrix</code>. We will add functionality to automatically threshold the gRNA UMI counts in a future update.</p>
<p>Also note that we say that a cell is <em>perturbed</em> by a gRNA if the corresponding thresholded indicator variable is 1, and <em>unperturbed</em> otherwise.</p>
</div>
<div id="select-genes-and-grnas-to-use-in-the-analysis" class="section level1">
<h1 class="hasAnchor">
<a href="#select-genes-and-grnas-to-use-in-the-analysis" class="anchor"></a>3. Select genes and gRNAs to use in the analysis</h1>
<p>We select the genes, gRNAs, and gene-gRNA pairs to use in the analysis. There are no hard-and-fast guidelines here, but as a starting point, we exclude genes that are expressed in fewer than 1,000 cells and gRNAs that are expressed in fewer than 100 cells. We ultimately keep 10,870 of the 32,738 genes and 1,560 of the 1,561 gRNAs.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># subset expressions matrix</span>
<span class="va">genes_to_keep</span> <span class="op">&lt;-</span> <span class="va">expressions</span><span class="op">@</span><span class="va">feature_covariates</span><span class="op">$</span><span class="va">n_nonzero</span> <span class="op">&gt;</span> <span class="fl">1000</span>
<span class="va">expressions_sub</span> <span class="op">&lt;-</span> <span class="va">expressions</span><span class="op">[</span><span class="va">genes_to_keep</span>,<span class="op">]</span>

<span class="co"># subset perturbations matrix</span>
<span class="va">gRNAs_to_keep</span> <span class="op">&lt;-</span> <span class="va">perturbations</span><span class="op">@</span><span class="va">feature_covariates</span><span class="op">$</span><span class="va">n_nonzero</span> <span class="op">&gt;</span> <span class="fl">100</span>
<span class="va">perturbations_sub</span> <span class="op">&lt;-</span> <span class="va">perturbations</span><span class="op">[</span><span class="va">gRNAs_to_keep</span>,<span class="op">]</span></code></pre></div>
<p>Next, we select the gene-gRNA pairs to analyze. The file <code>GSE120861_gene_gRNAgroup_pair_table.pilot.txt</code> contains a list of all the gene-gRNA pairs that were analyzed in the original study. The gRNAs are grouped into several categories, including non-targeting controls, transcription start sites, and DNase I hypersensitive sites (i.e., the candidate enhancer locations). First, we intersect the gene-gRNA pairs with the highly-expressed genes and gRNAs that we retained in the previous step. Next, we randomly sample 500 of the non-targeting control pairs for inclusion in the analysis. (This is for demonstration purposes only; if our goal were to assess calibration, we would want to sample the non-targeting controls in a more careful way.)</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># load the gene-gRNA pairs used in the original analysis</span>
<span class="va">original_pairs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_tsv</a></span><span class="op">(</span>file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">raw_data_dir</span>,
                                         <span class="st">"/GSE120861_gene_gRNAgroup_pair_table.pilot.txt"</span><span class="op">)</span>,
                           col_types <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"___cc___c____"</span><span class="op">)</span>, col_names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html">rename</a></span><span class="op">(</span><span class="st">"gRNA_id"</span> <span class="op">=</span> <span class="st">"gRNAgroup"</span>,
         <span class="st">"gene_id"</span> <span class="op">=</span> <span class="st">"ENSG.targetgene"</span>,
         <span class="st">"gRNA_type"</span> <span class="op">=</span> <span class="st">"general_group"</span><span class="op">)</span>

<span class="co"># get the IDs of the highly expressed genes and gRNAs</span>
<span class="va">genes_in_analysis</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get-names.html">get_feature_ids</a></span><span class="op">(</span><span class="va">expressions</span><span class="op">)</span>
<span class="va">gRNAs_in_analysis</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get-names.html">get_feature_ids</a></span><span class="op">(</span><span class="va">perturbations</span><span class="op">)</span>

<span class="co"># filter and sample</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">100</span><span class="op">)</span>
<span class="va">gene_gRNA_pairs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span>.data <span class="op">=</span> <span class="va">original_pairs</span>,
                          <span class="va">gRNA_id</span> <span class="op">%in%</span> <span class="va">gRNAs_in_analysis</span> <span class="op">&amp;&amp;</span> 
                            <span class="va">gene_id</span> <span class="op">%in%</span> <span class="va">genes_in_analysis</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">gRNA_type</span> <span class="op">%in%</span> <span class="st">"NTC"</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html">slice_sample</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">500</span><span class="op">)</span></code></pre></div>
<p>Finally, we save the gene-gRNA pairs that we have selected for analysis in the “processed data” directory.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">saveRDS</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">gene_gRNA_pairs</span>, file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">processed_data_dir</span>, <span class="st">"/gene_gRNA_pairs.rds"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div id="perform-quality-control-on-the-cells" class="section level1">
<h1 class="hasAnchor">
<a href="#perform-quality-control-on-the-cells" class="anchor"></a>4. Perform quality control on the cells</h1>
<p>The next step is to perform quality control on the cells. First, we create a <a href="https://timothy-barry.github.io/ondisc/articles/tutorial_other_classes.html">multimodal_ondisc_matrix</a>, combining the <code>expressions</code> and <code>perturbations</code> modalities into a single object.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">crispr_experiment</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/multimodal_ondisc_matrix.html">multimodal_ondisc_matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>expressions <span class="op">=</span> <span class="va">expressions_sub</span>,
                                                   perturbations <span class="op">=</span> <span class="va">perturbations_sub</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Using the raw, unfiltered global-cell covariate matrix, we exclude cells that received zero perturbations.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cells_to_keep</span> <span class="op">&lt;-</span> <span class="va">crispr_experiment</span><span class="op">@</span><span class="va">global_cell_covariates</span><span class="op">$</span><span class="va">perturbations_n_nonzero</span> <span class="op">&gt;</span> <span class="fl">0</span>
<span class="va">crispr_experiment_sub</span> <span class="op">&lt;-</span> <span class="va">crispr_experiment</span><span class="op">[</span>,<span class="va">cells_to_keep</span><span class="op">]</span></code></pre></div>
</div>
<div id="compute-the-cell-wise-covariate-matrix" class="section level1">
<h1 class="hasAnchor">
<a href="#compute-the-cell-wise-covariate-matrix" class="anchor"></a>5. Compute the cell-wise covariate matrix</h1>
<p>Finally, we compute the cell-wise covariate matrix. The following cell-specific covariates are reasonable choices to include in the analysis:</p>
<ol style="list-style-type: decimal">
<li>sequencing batch</li>
<li>percentage of UMIs that map to mitochondrial genes</li>
<li>log-transformed mRNA UMI count</li>
<li>log-transformed number of genes expressed in cell</li>
<li>log-transformed number of perturbations detected in cell, or log-transfored gRNA UMI count.</li>
</ol>
<p>In general, it probably is best to include as many of the above covariates in the analysis as possible. In our current example we can compute covariates 3-5:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cell_covariate_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarize</a></span><span class="op">(</span><span class="va">crispr_experiment_sub</span><span class="op">@</span><span class="va">global_cell_covariates</span>,
                                lg_umi_count <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log</a></span><span class="op">(</span><span class="va">expressions_n_umis</span><span class="op">)</span>,
                                lg_n_genes_expressed <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log</a></span><span class="op">(</span><span class="va">expressions_n_nonzero</span><span class="op">)</span>,
                                lg_n_perturbations <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log</a></span><span class="op">(</span><span class="va">perturbations_n_nonzero</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>We save the global cell-covariate matrix and the appropriately-subsetted gene and perturbation <code>ondisc_matrices</code> to the “processed data directory.”</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">saveRDS</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">cell_covariate_matrix</span>,
        file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">processed_data_dir</span>, <span class="st">"/cell_covariate_matrix.rds"</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">saveRDS</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">crispr_experiment_sub</span><span class="op">@</span><span class="va">modalities</span><span class="op">$</span><span class="va">expressions</span>,
        file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">processed_data_dir</span>, <span class="st">"/expressions.rds"</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">saveRDS</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">crispr_experiment_sub</span><span class="op">@</span><span class="va">modalities</span><span class="op">$</span><span class="va">perturbations</span>,
        file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">processed_data_dir</span>, <span class="st">"/perturbations.rds"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>The “processed data directory” should contain the following files:</p>
<ol style="list-style-type: decimal">
<li>The expressions.rds <code>ondisc_matrix</code> plus its backing .h5 file,</li>
<li>The perturbations.rds <code>ondisc_matsrix</code> plus its backing .h5 file,</li>
<li>The cell_covariate_matrix.rds file, and</li>
<li>The gene_gRNA_pairs.rds file.</li>
</ol>
<p>Now, we are ready to run our single-cell CRISPR screen analysis!</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Timothy Barry.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
