<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Tutorial 2: Using `metadata_ondisc_matrix` and `multimodal_ondisc_matrix` • ondisc</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/lumen/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Tutorial 2: Using `metadata_ondisc_matrix` and `multimodal_ondisc_matrix`">
<meta property="og:description" content="ondisc">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ondisc</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.9.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Function reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/tutorial_odm_class.html">Tutorial 1</a>
    </li>
    <li>
      <a href="../articles/tutorial_other_classes.html">Tutorial 2</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/Timothy-Barry/ondisc/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="tutorial_other_classes_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Tutorial 2: Using <code>metadata_ondisc_matrix</code> and <code>multimodal_ondisc_matrix</code>
</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/Timothy-Barry/ondisc/blob/master/vignettes/tutorial_other_classes.Rmd"><code>vignettes/tutorial_other_classes.Rmd</code></a></small>
      <div class="hidden name"><code>tutorial_other_classes.Rmd</code></div>

    </div>

    
    
<p>Tutorial 1 covered <code>ondisc_matrix</code>, the core class implemented by <code>ondisc</code>. Thus tutorial covers <code>metadata_ondisc_matrix</code> and <code>multimodal_ondisc_matrix</code>, two additional classes provided by the package. <code>metadata_ondisc_matrices</code> store cell-specific and feature-specific covariate matrices alongside the expression matrix, and <code>multimodal_ondisc_matrices</code> store multiple <code>metadata_ondisc_matrices</code> representing different cellular modalities. Together, <code>metadata_ondisc_matrix</code> and <code>multimodal_ondisc_matrix</code> facilitate feature selection, quality control, and other common single-cell data preprocessing tasks.</p>
<p>We begin by loading the package.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Timothy-Barry/ondisc">ondisc</a></span><span class="op">)</span></code></pre></div>
<div id="the-metadata_ondisc_matrix-class" class="section level1">
<h1 class="hasAnchor">
<a href="#the-metadata_ondisc_matrix-class" class="anchor"></a>The <code>metadata_ondisc_matrix</code> class</h1>
<p>A <code>metadata_ondisc_matrix</code> object consists of three components: (i) an <code>ondisc_matrix</code> representing the expression data, (ii) a data frame storing the cell-specific covariates, and (iii) a data frame storing the feature-specific covariates. The easiest way to initialize a <code>metadata_ondisc_matrix</code> is by calling <code>create_ondisc_matrix_from_mtx</code> on an mtx file and associated metadata files, setting the optional parameter <code>return_metadata_ondisc_matrix</code> to <code>TRUE</code>. Below, we reproduce the example from Tutorial 1, this time returning a <code>metadata_ondisc_matrix</code> instead of a list.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a><span class="co"># Set paths to the .mtx and .tsv files</span></span>
<span id="cb2-2"><a href="#cb2-2"></a>raw_data_dir &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"extdata"</span>, <span class="dt">package =</span> <span class="st">"ondisc"</span>)</span>
<span id="cb2-3"><a href="#cb2-3"></a>mtx_fp &lt;-<span class="st"> </span><span class="kw">paste0</span>(raw_data_dir, <span class="st">"/gene_expression.mtx"</span>)</span>
<span id="cb2-4"><a href="#cb2-4"></a>barcodes_fp &lt;-<span class="st"> </span><span class="kw">paste0</span>(raw_data_dir, <span class="st">"/cell_barcodes.tsv"</span>)</span>
<span id="cb2-5"><a href="#cb2-5"></a>features_fp &lt;-<span class="st"> </span><span class="kw">paste0</span>(raw_data_dir, <span class="st">"/genes.tsv"</span>)</span>
<span id="cb2-6"><a href="#cb2-6"></a></span>
<span id="cb2-7"><a href="#cb2-7"></a><span class="co"># Specify directory in which to store the .h5 file</span></span>
<span id="cb2-8"><a href="#cb2-8"></a>temp_dir &lt;-<span class="st"> </span><span class="kw">tempdir</span>()</span>
<span id="cb2-9"><a href="#cb2-9"></a></span>
<span id="cb2-10"><a href="#cb2-10"></a><span class="co"># Initialize metadata_ondisc_matrix</span></span>
<span id="cb2-11"><a href="#cb2-11"></a>expressions &lt;-<span class="st"> </span><span class="kw">create_ondisc_matrix_from_mtx</span>(<span class="dt">mtx_fp =</span> mtx_fp,</span>
<span id="cb2-12"><a href="#cb2-12"></a>                                              <span class="dt">barcodes_fp =</span> barcodes_fp,</span>
<span id="cb2-13"><a href="#cb2-13"></a>                                              <span class="dt">features_fp =</span> features_fp,</span>
<span id="cb2-14"><a href="#cb2-14"></a>                                              <span class="dt">on_disk_dir =</span> temp_dir,</span>
<span id="cb2-15"><a href="#cb2-15"></a>                                              <span class="dt">return_metadata_ondisc_matrix =</span> <span class="ot">TRUE</span>)</span>
<span id="cb2-16"><a href="#cb2-16"></a><span class="co">#&gt; </span></span>
<span id="cb2-17"><a href="#cb2-17"></a><span class="op">|</span><span class="er">========</span><span class="st">                                                                 </span><span class="er">|</span><span class="st"> </span><span class="dv">11</span>%</span>
<span id="cb2-18"><a href="#cb2-18"></a><span class="op">|</span><span class="er">=================</span><span class="st">                                                        </span><span class="er">|</span><span class="st"> </span><span class="dv">23</span>%</span>
<span id="cb2-19"><a href="#cb2-19"></a><span class="op">|</span><span class="er">==========================</span><span class="st">                                               </span><span class="er">|</span><span class="st"> </span><span class="dv">36</span>%</span>
<span id="cb2-20"><a href="#cb2-20"></a><span class="op">|</span><span class="er">====================================</span><span class="st">                                     </span><span class="er">|</span><span class="st"> </span><span class="dv">48</span>%</span>
<span id="cb2-21"><a href="#cb2-21"></a><span class="op">|</span><span class="er">=============================================</span><span class="st">                            </span><span class="er">|</span><span class="st"> </span><span class="dv">61</span>%</span>
<span id="cb2-22"><a href="#cb2-22"></a><span class="op">|</span><span class="er">======================================================</span><span class="st">                   </span><span class="er">|</span><span class="st"> </span><span class="dv">73</span>%</span>
<span id="cb2-23"><a href="#cb2-23"></a><span class="op">|</span><span class="er">===============================================================</span><span class="st">          </span><span class="er">|</span><span class="st"> </span><span class="dv">86</span>%</span>
<span id="cb2-24"><a href="#cb2-24"></a><span class="op">|</span><span class="er">=========================================================================|</span><span class="st"> </span><span class="dv">98</span>%</span>
<span id="cb2-25"><a href="#cb2-25"></a><span class="op">|</span><span class="er">=========================================================================|</span><span class="st"> </span><span class="dv">100</span>%</span>
<span id="cb2-26"><a href="#cb2-26"></a><span class="co">#&gt; Writing CSC data.</span></span>
<span id="cb2-27"><a href="#cb2-27"></a><span class="co">#&gt; Writing CSR data.</span></span></code></pre></div>
<p>The variable <code>expressions</code> is an object of class <code>metadata_ondisc_matrix</code>; <code>expressions</code> contains the fields <code>ondisc_matrix</code>, <code>cell_covariates</code>, and <code>feature_covariates</code>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Print the variable</span>
<span class="va">expressions</span>
<span class="co">#&gt; A metadata_ondisc_matrix with the following components:</span>
<span class="co">#&gt;  An ondisc_matrix with 300 features and 900 cells.</span>
<span class="co">#&gt;  A cell covariate matrix with columns n_nonzero, n_umis, p_mito.</span>
<span class="co">#&gt;  A feature covariate matrix with columns mean_expression, coef_of_variation, n_nonzero.</span></code></pre></div>
<p>We alternately can initialize a <code>metadata_ondisc_matrix</code> by calling the constructor function of the <code>metadata_ondisc_matrix</code> class; see ?metadata_ondisc_matrix for details.</p>
</div>
<div id="the-multimodal_ondisc_matrix-class" class="section level1">
<h1 class="hasAnchor">
<a href="#the-multimodal_ondisc_matrix-class" class="anchor"></a>The <code>multimodal_ondisc_matrix</code> class</h1>
<p>The <code>multimodal_ondisc_matrix</code> class is used to represent multimodal data. <code>multimodal_ondisc_matrix</code> objects have two fields: (i) a named list of <code>metadata_ondisc_matrices</code> representing different modalities, and (ii) a global (i.e., cross-modality) cell-specific covariate matrix. The <code>ondisc</code> package ships with example CRISPR perturbation data, which we use to initialize a new perturbation modality via a call to <code>create_ondisc_matrix_from_mtx</code>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Set paths to the perturbation .mtx and .tsv files</span>
<span class="va">mtx_fp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">raw_data_dir</span>, <span class="st">"/perturbation.mtx"</span><span class="op">)</span>
<span class="va">barcodes_fp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">raw_data_dir</span>, <span class="st">"/cell_barcodes.tsv"</span><span class="op">)</span>
<span class="va">features_fp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">raw_data_dir</span>, <span class="st">"/guides.tsv"</span><span class="op">)</span>

<span class="co"># Initialize metadata_ondisc_matrix</span>
<span class="va">perturbations</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_ondisc_matrix_from_mtx.html">create_ondisc_matrix_from_mtx</a></span><span class="op">(</span>mtx_fp <span class="op">=</span> <span class="va">mtx_fp</span>,
                                               barcodes_fp <span class="op">=</span> <span class="va">barcodes_fp</span>,
                                               features_fp <span class="op">=</span> <span class="va">features_fp</span>,
                                               on_disk_dir <span class="op">=</span> <span class="va">temp_dir</span>,
                                               return_metadata_ondisc_matrix <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Writing CSC data.</span>
<span class="co">#&gt; Writing CSR data.</span></code></pre></div>
<p>Like <code>expressions</code>, the variable <code>perturbations</code> is an object of class <code>metadata_ondisc_matrix</code>. However, because <code>perturbations</code> represents logical perturabtion data instead of integer gene expression data, the cell-specific and feature-specific covariates of <code>perturbations</code> differ from those of <code>expressions</code>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># These matrices have different columns</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">expressions</span><span class="op">@</span><span class="va">cell_covariates</span><span class="op">)</span>
<span class="co">#&gt;   n_nonzero n_umis     p_mito</span>
<span class="co">#&gt; 1        43    214 0.04672897</span>
<span class="co">#&gt; 2        26    169 0.00000000</span>
<span class="co">#&gt; 3        22    116 0.05172414</span>
<span class="co">#&gt; 4        37    258 0.08139535</span>
<span class="co">#&gt; 5        36    224 0.08035714</span>
<span class="co">#&gt; 6        31    147 0.07482993</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">perturbations</span><span class="op">@</span><span class="va">cell_covariates</span><span class="op">)</span>
<span class="co">#&gt;   n_nonzero</span>
<span class="co">#&gt; 1        21</span>
<span class="co">#&gt; 2        11</span>
<span class="co">#&gt; 3        17</span>
<span class="co">#&gt; 4        14</span>
<span class="co">#&gt; 5         7</span>
<span class="co">#&gt; 6        11</span></code></pre></div>
<p>The <code>expressions</code> and <code>perturbations</code> data are multimodal – they are collected from the same set of cells. We can create a <code>multimodal_ondisc_matrix</code> by passing a named list of <code>metadata_ondisc_matrix</code> objects – in this case, <code>expressions</code> and <code>perturbations</code> – to the constructor function of the <code>multimodal_ondisc_matrix</code> class.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">modality_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>expressions <span class="op">=</span> <span class="va">expressions</span>, perturbations <span class="op">=</span> <span class="va">perturbations</span><span class="op">)</span>
<span class="va">crispr_experiment</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/multimodal_ondisc_matrix.html">multimodal_ondisc_matrix</a></span><span class="op">(</span><span class="va">modality_list</span><span class="op">)</span></code></pre></div>
<p>The variable <code>crispr_experiment</code> is an object of class <code>multimodal_ondisc_matrix</code>. The column names of the global covariate matrix are derived from the names of the modalities.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># print variable</span>
<span class="va">crispr_experiment</span>
<span class="co">#&gt; A multimodal_ondisc_matrix with the following modalities:</span>
<span class="co">#&gt; 1. expressions A metadata_ondisc_matrix with the following components:</span>
<span class="co">#&gt;  An ondisc_matrix with 300 features and 900 cells.</span>
<span class="co">#&gt;  A cell covariate matrix with columns n_nonzero, n_umis, p_mito.</span>
<span class="co">#&gt;  A feature covariate matrix with columns mean_expression, coef_of_variation, n_nonzero.</span>
<span class="co">#&gt; 2. perturbations A metadata_ondisc_matrix with the following components:</span>
<span class="co">#&gt;  An ondisc_matrix with 250 features and 900 cells.</span>
<span class="co">#&gt;  A cell covariate matrix with columns n_nonzero.</span>
<span class="co">#&gt;  A feature covariate matrix with columns n_nonzero.</span>
<span class="co">#&gt; Plus, a global cell-covariate matrix containing 4 covariates and 900 cells.</span>

<span class="co"># show the global covariate matrix</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">crispr_experiment</span><span class="op">@</span><span class="va">global_cell_covariates</span><span class="op">)</span>
<span class="co">#&gt;   expressions_n_nonzero expressions_n_umis expressions_p_mito</span>
<span class="co">#&gt; 1                    43                214         0.04672897</span>
<span class="co">#&gt; 2                    26                169         0.00000000</span>
<span class="co">#&gt; 3                    22                116         0.05172414</span>
<span class="co">#&gt; 4                    37                258         0.08139535</span>
<span class="co">#&gt; 5                    36                224         0.08035714</span>
<span class="co">#&gt; 6                    31                147         0.07482993</span>
<span class="co">#&gt;   perturbations_n_nonzero</span>
<span class="co">#&gt; 1                      21</span>
<span class="co">#&gt; 2                      11</span>
<span class="co">#&gt; 3                      17</span>
<span class="co">#&gt; 4                      14</span>
<span class="co">#&gt; 5                       7</span>
<span class="co">#&gt; 6                      11</span></code></pre></div>
<p>The figure below summarizes the relationship between <code>ondisc_matrix</code>, <code>metadata_ondisc_matrix</code>, and <code>multimodal_ondisc_matrix</code>.</p>
<div class="figure">
<img src="classes_cropped.jpg" alt="**Figure**: Classes provided by the package. a) `ondisc_matrix`, b) `metadata_ondisc_matrix`, c) `multimodal_ondisc_matrix` " width="55%"><p class="caption">
<strong>Figure</strong>: Classes provided by the package. a) <code>ondisc_matrix</code>, b) <code>metadata_ondisc_matrix</code>, c) <code>multimodal_ondisc_matrix</code>
</p>
</div>
</div>
<div id="querying-basic-information" class="section level1">
<h1 class="hasAnchor">
<a href="#querying-basic-information" class="anchor"></a>Querying basic information</h1>
<p>We can use the functions <code>get_feature_ids</code>, <code>get_feature_names</code>, and <code>get_cell_barcodes</code> to obtain the feature IDs, feature names (if applicable), and cell barcodes, respectively, of a <code>metadata_ondisc_matrix</code> or a <code>multimodal_ondisc_matrix</code>. <code>get_feature_ids</code> and <code>get_feature_names</code> return a list when called on a <code>multimodal_ondisc_matrix</code>, as the different modalities contain different features.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># metadata_ondisc_matrix</span>
<span class="va">cell_barcodes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get-names.html">get_cell_barcodes</a></span><span class="op">(</span><span class="va">expressions</span><span class="op">)</span>
<span class="va">feature_ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get-names.html">get_feature_ids</a></span><span class="op">(</span><span class="va">expressions</span><span class="op">)</span>
<span class="va">feature_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get-names.html">get_feature_names</a></span><span class="op">(</span><span class="va">expressions</span><span class="op">)</span>

<span class="co"># multimodal_ondisc_matrix</span>
<span class="va">cell_barcodes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get-names.html">get_cell_barcodes</a></span><span class="op">(</span><span class="va">crispr_experiment</span><span class="op">)</span>
<span class="va">feature_ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get-names.html">get_feature_ids</a></span><span class="op">(</span><span class="va">crispr_experiment</span><span class="op">)</span></code></pre></div>
<p>We likewise can use <code>dim</code>, <code>nrow</code>, and <code>ncol</code> to query the dimension, number of rows, and number of columns of a <code>metadata_ondisc_matrix</code> or <code>multimodal_ondisc_matrix</code>. <code>dim</code> and <code>nrow</code> return lists when called on a <code>multimodal_ondisc_matrix</code>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># metadata_ondisc_matrix</span>
<span class="fu"><a href="../reference/dim.html">dim</a></span><span class="op">(</span><span class="va">expressions</span><span class="op">)</span>
<span class="co">#&gt; [1] 300 900</span>
<span class="fu"><a href="../reference/dim.html">nrow</a></span><span class="op">(</span><span class="va">expressions</span><span class="op">)</span>
<span class="co">#&gt; [1] 300</span>
<span class="fu"><a href="../reference/dim.html">ncol</a></span><span class="op">(</span><span class="va">expressions</span><span class="op">)</span>
<span class="co">#&gt; [1] 900</span>

<span class="co"># multimodal_ondisc_matrix</span>
<span class="fu"><a href="../reference/dim.html">dim</a></span><span class="op">(</span><span class="va">crispr_experiment</span><span class="op">)</span>
<span class="co">#&gt; $expressions</span>
<span class="co">#&gt; [1] 300 900</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $perturbations</span>
<span class="co">#&gt; [1] 250 900</span>
<span class="fu"><a href="../reference/dim.html">nrow</a></span><span class="op">(</span><span class="va">crispr_experiment</span><span class="op">)</span>
<span class="co">#&gt; $expressions</span>
<span class="co">#&gt; [1] 300</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $perturbations</span>
<span class="co">#&gt; [1] 250</span>
<span class="fu"><a href="../reference/dim.html">ncol</a></span><span class="op">(</span><span class="va">crispr_experiment</span><span class="op">)</span>
<span class="co">#&gt; [1] 900</span></code></pre></div>
</div>
<div id="subsetting" class="section level1">
<h1 class="hasAnchor">
<a href="#subsetting" class="anchor"></a>Subsetting</h1>
<p>Similar to <code>ondisc_matrices</code>, <code>metadata_ondisc_matrices</code> and <code>multimodal_ondisc_matrices</code> can be subset using the <code><a href="https://rdrr.io/r/base/Extract.html">[</a></code> operator. <code>metadata_ondisc_matrices</code> can be subset either by feature or cell, while <code>multimodal_ondisc_matrices</code> can be subset by cell only.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># metadata_ondisc_matrix</span>
<span class="co"># keep cells 100 - 150</span>
<span class="va">expressions_sub</span> <span class="op">&lt;-</span> <span class="va">expressions</span><span class="op">[</span>,<span class="fl">100</span><span class="op">:</span><span class="fl">150</span><span class="op">]</span>
<span class="co"># keep genes ENSG00000188305, ENSG00000257284, ENSG00000251655</span>
<span class="va">expressions_sub</span> <span class="op">&lt;-</span> <span class="va">expressions</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"ENSG00000188305"</span>, <span class="st">"ENSG00000257284"</span>, <span class="st">"ENSG00000251655"</span><span class="op">)</span>,<span class="op">]</span>

<span class="co"># multimodal_ondisc_matrix</span>
<span class="co"># keep all cells except 1 - 100</span>
<span class="va">crispr_experiment_sub</span> <span class="op">&lt;-</span> <span class="va">crispr_experiment</span><span class="op">[</span>,<span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">100</span><span class="op">)</span><span class="op">]</span></code></pre></div>
<p>As with <code>ondisc_matrices</code>, the original objects remain unchanged.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">expressions</span>
<span class="co">#&gt; A metadata_ondisc_matrix with the following components:</span>
<span class="co">#&gt;  An ondisc_matrix with 300 features and 900 cells.</span>
<span class="co">#&gt;  A cell covariate matrix with columns n_nonzero, n_umis, p_mito.</span>
<span class="co">#&gt;  A feature covariate matrix with columns mean_expression, coef_of_variation, n_nonzero.</span>
<span class="va">crispr_experiment</span>
<span class="co">#&gt; A multimodal_ondisc_matrix with the following modalities:</span>
<span class="co">#&gt; 1. expressions A metadata_ondisc_matrix with the following components:</span>
<span class="co">#&gt;  An ondisc_matrix with 300 features and 900 cells.</span>
<span class="co">#&gt;  A cell covariate matrix with columns n_nonzero, n_umis, p_mito.</span>
<span class="co">#&gt;  A feature covariate matrix with columns mean_expression, coef_of_variation, n_nonzero.</span>
<span class="co">#&gt; 2. perturbations A metadata_ondisc_matrix with the following components:</span>
<span class="co">#&gt;  An ondisc_matrix with 250 features and 900 cells.</span>
<span class="co">#&gt;  A cell covariate matrix with columns n_nonzero.</span>
<span class="co">#&gt;  A feature covariate matrix with columns n_nonzero.</span>
<span class="co">#&gt; Plus, a global cell-covariate matrix containing 4 covariates and 900 cells.</span></code></pre></div>
</div>
<div id="notes-and-tips" class="section level1">
<h1 class="hasAnchor">
<a href="#notes-and-tips" class="anchor"></a>Notes and tips</h1>
<ul>
<li><p>To save and load a <code>metadata_ondisc_matrix</code> or <code>multimodal_ondisc_matrix</code>, simply use the functions <code>saveRDS</code> and <code>readRDS</code>, respectively.</p></li>
<li><p>It is not possible to apply the double-bracket “pull submatrix into memory” operator to a <code>metadata_ondisc_matrix</code> or <code>multimodal_ondisc_matrix</code>. To apply the double-bracket operator to an <code>ondisc_matrix</code> stored <em>inside</em> a <code>metadata_ondisc_matrix</code> or <code>multimodal_ondisc_matrix</code>, simply access the <code>ondisc_matrix</code> first.</p></li>
<li><p>The <code>metadata_ondisc_matrix</code> and <code>multimodal_ondisc_matrix</code> classes are inspired by analogous classes in the <code>SingleCellExperiment</code> and <code>Seurat</code> packages. However, unlike <code>SingleCellExperiment</code> and <code>Seurat</code>, <code>ondisc</code> does not (currently) support storing the dense matrix of normalized UMI counts. We view this as a feature, not a bug. Storing the normalized expression matrix is almost never necessary and is hugely expensive (in space and time) for large-scale data.</p></li>
</ul>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://timothy-barry.github.io/">Tim Barry</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
