<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Using ondisc • ondisc</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/lumen/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Using ondisc">
<meta property="og:description" content="ondisc">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ondisc</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.9.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Function reference</a>
</li>
<li>
  <a href="../articles/using_ondisc_static.html">Basic usage</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/Timothy-Barry/ondisc/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="using_ondisc_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Using ondisc</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/Timothy-Barry/ondisc/blob/master/vignettes/using_ondisc.Rmd"><code>vignettes/using_ondisc.Rmd</code></a></small>
      <div class="hidden name"><code>using_ondisc.Rmd</code></div>

    </div>

    
    
<p>This vignette runs through an example of using the <code>ondisc</code> package. First, we download an example single-cell dataset from 10X Genomics. Next, we store the data in an <code>on_disc_matrix</code>, the core object provided by <code>ondisc</code>. Finally, we perform some simple operations on the initialized <code>on_disc_matrix</code>, including subset, pull sub-matrix into memory, and compute summary statistics. Let’s load the required packages for this vignette.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/HenrikBengtsson/R.utils">R.utils</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Timothy-Barry/ondisc">ondisc</a></span><span class="op">)</span></code></pre></div>
<p>We work with the public <a href="https://support.10xgenomics.com/single-cell-gene-expression/datasets/4.0.0/Parent_NGSC3_DI_PBMC">PBMC dataset</a> published by 10X Genomics. We download the <code>matrix.mtx</code>, <code>barcodes.tsv</code>, and <code>features.tsv</code> files from the 10X website into a specified directory, <code>data_dir</code>. The files are 88 MB in total, so the download should not take too long. Be sure to change the file path of <code>data_dir</code> to a location on your machine.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">data_dir</span> <span class="op">&lt;-</span> <span class="st">"/Users/timbarry/Box/onDisc_all/onDisc_offsite/raw_data"</span> <span class="co"># change this file path!</span>
<span class="fu"><a href="https://rdrr.io/r/utils/download.file.html">download.file</a></span><span class="op">(</span>url <span class="op">=</span> <span class="st">"https://cf.10xgenomics.com/samples/cell-exp/4.0.0/Parent_NGSC3_DI_PBMC/Parent_NGSC3_DI_PBMC_filtered_feature_bc_matrix.tar.gz"</span>, destfile <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">data_dir</span>, <span class="st">"/pbmc.tar.gz"</span><span class="op">)</span><span class="op">)</span>
<span class="va">fs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"filtered_feature_bc_matrix/"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"features.tsv.gz"</span>, <span class="st">"matrix.mtx.gz"</span>, <span class="st">"barcodes.tsv.gz"</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/untar.html">untar</a></span><span class="op">(</span>tarfile <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">data_dir</span>, <span class="st">"/pbmc.tar.gz"</span><span class="op">)</span>, files <span class="op">=</span> <span class="va">fs</span>, exdir <span class="op">=</span> <span class="va">data_dir</span><span class="op">)</span>
<span class="kw">for</span> <span class="op">(</span><span class="va">curr_f</span> <span class="kw">in</span> <span class="va">fs</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/pkg/R.utils/man/compressFile.html">gunzip</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">data_dir</span>, <span class="st">"/"</span>, <span class="va">curr_f</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>If the above code chunk worked, then the downloaded files are unzipped and stored in the subdirectory <code>filtered_feature_bc_matrix</code>. Let’s list the files now and save the file paths to variables.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">data_sub_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">data_dir</span>, <span class="st">"/filtered_feature_bc_matrix"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span><span class="op">(</span><span class="va">data_sub_dir</span><span class="op">)</span>
<span class="va">mtx_fp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">data_sub_dir</span>, <span class="st">"/matrix.mtx"</span><span class="op">)</span>
<span class="va">barcode_fp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">data_sub_dir</span>, <span class="st">"/barcodes.tsv"</span><span class="op">)</span>
<span class="va">features_fp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">data_sub_dir</span>, <span class="st">"/features.tsv"</span><span class="op">)</span></code></pre></div>
<p>If you would like to use a small, example dataset that ships with the package instead of the 10X PBMC data, execute the following chunk.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># change to TRUE if you instead would like to use a small, pre-packaged dataset!</span>
  <span class="va">data_sub_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/find.package.html">find.package</a></span><span class="op">(</span>package <span class="op">=</span> <span class="st">"ondisc"</span><span class="op">)</span>, <span class="st">"/extdata"</span><span class="op">)</span>
  <span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span><span class="op">(</span><span class="va">data_sub_dir</span><span class="op">)</span>
  <span class="va">mtx_fp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">data_sub_dir</span>, <span class="st">"/matrix_1.mtx"</span><span class="op">)</span>
  <span class="va">barcode_fp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">data_sub_dir</span>, <span class="st">"/barcodes_1.tsv"</span><span class="op">)</span>
  <span class="va">features_fp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">data_sub_dir</span>, <span class="st">"/features_1.tsv"</span><span class="op">)</span>
<span class="op">}</span></code></pre></div>
<div id="initialize-an-on_disc_matrix" class="section level2">
<h2 class="hasAnchor">
<a href="#initialize-an-on_disc_matrix" class="anchor"></a>Initialize an <code>on_disc_matrix</code>
</h2>
<p>Next, we initialize an <code>on_disc_matrix</code> that will contain the expression data. An <code>on_disc_matrix</code> is the core object implemented by <code>ondisc</code>; its purpose is to store a gene-by-cell expression matrix on-disk rather than in-memory, facilitating large-scale computations on expression matrices. We initialize an <code>on_disc_matrix</code> from a <code>.mtx</code> file using the function <code>create_on_disc_matrix_from_10x_mtx</code>. We pass the file paths of <code>matrix.mtx</code>, <code>barcodes.tsv</code>, and <code>features.tsv</code> as arguments to the function, as well as the on-disk directory in which we would like to store the expression data.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">on_disc_dir</span> <span class="op">&lt;-</span> <span class="va">data_sub_dir</span>
<span class="va">expression_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_on_disc_matrix_from_10x_mtx.html">create_on_disc_matrix_from_10x_mtx</a></span><span class="op">(</span>mtx_fp <span class="op">=</span> <span class="va">mtx_fp</span>, barcode_fp <span class="op">=</span> <span class="va">barcode_fp</span>, features_fp <span class="op">=</span> <span class="va">features_fp</span>, on_disc_dir <span class="op">=</span> <span class="va">on_disc_dir</span><span class="op">)</span></code></pre></div>
<p>The function <code>create_on_disc_matrix_from_10x_mtx</code> reads the expression data from the <code>.mtx</code> and <code>.tsv</code> files, manipulates the data, and then stores the data in a specially-formatted HDF5 file named <code>on_disc_matrix_1.h5</code>. (If there already exists a file named <code>on_disc_matrix_1.h5</code> in the specified directory, then <code>create_on_disc_matrix_from_10x_mtx</code> instead stores the expression data in a file called <code>on_disc_matrix_2.h5</code>, and so on.) <code>create_on_disc_matrix_from_10x_mtx</code> returns an R object of class <code>on_disc_matrix</code>, which in the above example, we save to the variable <code>expression_matrix</code>. An <code>on_disc_matrix</code> is an S4 object that contains three slots: a path to the on-disk HDF5 file that stores the expression data, an integer vector storing the cells currently in use, and an integer vector storing the genes currently in use. Thus, <code>on_disc_matrix</code> objects typically have quite a small memory footprint (in this example, only about 1.2 KB or 1200 bytes).</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">expression_matrix</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/object.size.html">object.size</a></span><span class="op">(</span><span class="va">expression_matrix</span><span class="op">)</span></code></pre></div>
<p>The algorithm underlying <code>create_on_disc_matrix_from_10x_mtx</code> is highly efficient. However, <code>on_disc_matrices</code> are stored as HDF5 files under the hood, and because HDF5 does not currently support multithreaded write, <code>create_on_disc_matrix_from_10x_mtx</code> is in practice slower than one might expect. The code in the above chunk takes about fifteen minutes to run on the real data and a few seconds to run on the in-package example data. However, <code>create_on_disc_matrix_from_10x_mtx</code> only needs to be run once per dataset, even after closing and starting new R sessions. Additionally, the designers of HDF5 are <a href="https://forum.hdfgroup.org/t/multi-threading-in-the-hdf5-library/6462">considering</a> implementing multithreaded read and write. <code>create_on_disc_matrix_from_10x_mtx</code> will become faster if and when this feature becomes available.</p>
<p>It is also possible to initialize an <code>on_disc_matrix</code> from an .h5 file or an in-memory R matrix. See the article XXX.</p>
</div>
<div id="obtain-basic-information-from-an-on_disc_matrix" class="section level2">
<h2 class="hasAnchor">
<a href="#obtain-basic-information-from-an-on_disc_matrix" class="anchor"></a>Obtain basic information from an <code>on_disc_matrix</code>
</h2>
<p>The newly-created variable <code>expression_matrix</code> is an object of class <code>on_disc_matrix</code>. Evaluating <code>expression_matrix</code> in the console prints its class, as well as the number of genes and cells in the matrix.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">expression_matrix</span></code></pre></div>
<p>Calling <code>head</code> prints the expressions of the first few genes and cells.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">expression_matrix</span><span class="op">)</span></code></pre></div>
<p>The functions <code>get_gene_ids</code>, <code>get_gene_names</code>, and <code>get_cell_barcodes</code> return the gene IDs, (human-readable) gene names, and cell barcodes, respectively.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gene_ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_cell_barcodes.html">get_gene_ids</a></span><span class="op">(</span><span class="va">expression_matrix</span><span class="op">)</span>
<span class="va">gene_ids</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>
<span class="va">gene_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_cell_barcodes.html">get_gene_names</a></span><span class="op">(</span><span class="va">expression_matrix</span><span class="op">)</span>
<span class="va">gene_names</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>
<span class="va">cell_barcodes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_cell_barcodes.html">get_cell_barcodes</a></span><span class="op">(</span><span class="va">expression_matrix</span><span class="op">)</span>
<span class="va">cell_barcodes</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></code></pre></div>
<p>Finally, <code>dim</code>, <code>nrow</code>, and <code>ncol</code> return the dimension, number of rows (i.e., number of genes), and number of columns (i.e., number of cells) of the expression matrix.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">expression_matrix</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">expression_matrix</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">expression_matrix</span><span class="op">)</span></code></pre></div>
</div>
<div id="subset" class="section level2">
<h2 class="hasAnchor">
<a href="#subset" class="anchor"></a>Subset</h2>
<p>We can subset an <code>on_disc_matrix</code> to obtain a new <code>on_disc_matrix</code> that is a submatrix of the original. Subsetting <code>on_disc_matrices</code> is very similar to subsetting normal R matrices. We use single square brackets and specify the rows and columns that we would like to keep by integer position.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">x</span> <span class="op">&lt;-</span> <span class="va">expression_matrix</span><span class="op">[</span><span class="fl">100</span><span class="op">:</span><span class="fl">110</span>,<span class="op">]</span> <span class="co"># keep genes 100-110</span>
<span class="va">x</span> <span class="op">&lt;-</span> <span class="va">expression_matrix</span><span class="op">[</span>,<span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">10</span>,<span class="fl">20</span><span class="op">)</span><span class="op">]</span> <span class="co"># keep all cells except cells 10 and 20</span>
<span class="va">x</span> <span class="op">&lt;-</span> <span class="va">expression_matrix</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">100</span>, <span class="fl">200</span><span class="op">:</span><span class="fl">250</span><span class="op">)</span>, <span class="fl">300</span><span class="op">:</span><span class="fl">500</span><span class="op">]</span> <span class="co"># keep genes 1-100, 200-250 and cells 300-500</span>
<span class="va">x</span> <span class="co"># x has class on_disc_matrix, just like expression_matrix</span></code></pre></div>
<p>We also can subset <code>on_disc_matrices</code> using logical or character vectors. Character vectors correspond to <em>gene ID</em> and <em>cell barcode</em>, respectively.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">x</span> <span class="op">&lt;-</span> <span class="va">expression_matrix</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"ENSG00000260528"</span>, <span class="st">"ENSG00000258908"</span>, <span class="st">"ENSG00000081913"</span><span class="op">)</span>,<span class="op">]</span> <span class="co"># keep these three genes</span>
<span class="va">x</span> <span class="op">&lt;-</span> <span class="va">expression_matrix</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"ACAGCCGCAGAAACCG"</span>, <span class="st">"CTACTATAGTGTACCT"</span><span class="op">)</span><span class="op">]</span> <span class="co"># keep these two cells</span>
<span class="va">x</span> <span class="op">&lt;-</span> <span class="va">expression_matrix</span><span class="op">[</span><span class="op">!</span><span class="op">(</span><span class="fu"><a href="../reference/get_cell_barcodes.html">get_gene_ids</a></span><span class="op">(</span><span class="va">expression_matrix</span><span class="op">)</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"ENSG00000167525"</span>, <span class="st">"ENSG00000235815"</span><span class="op">)</span><span class="op">)</span>,<span class="op">]</span> <span class="co"># keep all genes except ENSG00000167525 and ENSG00000235815</span></code></pre></div>
<p>Subsetting an <code>on_disc_matrix</code> leaves the original object unchanged.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">expression_matrix</span> <span class="co"># unchanged</span></code></pre></div>
<p>This important property, called <em>object persistence</em>, makes programming with <code>on_disc_matrices</code> easier and more intuitive.</p>
</div>
<div id="pull-submatrix-into-memory" class="section level2">
<h2 class="hasAnchor">
<a href="#pull-submatrix-into-memory" class="anchor"></a>Pull submatrix into memory</h2>
<p>We can pull submatrices of an <code>on_disc_matrix</code> into memory, allowing us to perform operations on subsets of the data, like regressing on a subset of genes or examining the distribution of a subset of cells. To pull a submatrix into memory, specify the genes and/or cells to extract using <em>double</em> (as opposed to <em>single</em>) square brackets.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">m</span> <span class="op">&lt;-</span> <span class="va">expression_matrix</span><span class="op">[[</span><span class="fl">6</span>, <span class="op">]</span><span class="op">]</span> <span class="co"># pull gene 6 into memory</span>
<span class="va">m</span> <span class="op">&lt;-</span> <span class="va">expression_matrix</span><span class="op">[[</span>,<span class="fl">100</span><span class="op">]</span><span class="op">]</span> <span class="co"># pull cell 100 into memory</span>
<span class="va">m</span> <span class="op">&lt;-</span> <span class="va">expression_matrix</span><span class="op">[[</span><span class="fl">100</span><span class="op">:</span><span class="fl">110</span>, <span class="fl">200</span><span class="op">:</span><span class="fl">250</span><span class="op">]</span><span class="op">]</span> <span class="co"># pull genes 100-110 and cells 200-250 into memory</span></code></pre></div>
<p>The output of a double-bracket “pull into-memory” operation is a sparse matrix, implemented by the package <code>Matrix</code>. <code>Matrix</code> provides many useful functions for working with sparse matrices, in addition to all the standard matrix operations.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span><span class="op">(</span><span class="va">m</span><span class="op">)</span> <span class="co"># m is of class Matrix</span></code></pre></div>
<p>Like with subsetting, we can pull submatrices into memory using a character vector or logical vector instead of a numeric vector.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">m</span> <span class="op">&lt;-</span> <span class="va">expression_matrix</span><span class="op">[[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"ENSG00000260528"</span>, <span class="st">"ENSG00000258908"</span>, <span class="st">"ENSG00000081913"</span><span class="op">)</span>,<span class="op">]</span><span class="op">]</span> <span class="co"># pull genes ENSG00000260528, ENSG00000258908, and ENSG00000081913 into memory</span>
<span class="va">m</span> <span class="op">&lt;-</span> <span class="va">expression_matrix</span><span class="op">[[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"ACAGCCGCAGAAACCG"</span>, <span class="st">"CTACTATAGTGTACCT"</span><span class="op">)</span><span class="op">]</span><span class="op">]</span> <span class="co"># pull cells ACAGCCGCAGAAACCG and CTACTATAGTGTACCT into memory</span></code></pre></div>
<p>Recall that, when working with lists, single brackets return a subsetted list and double brackets return an object stored <em>inside</em> the list. The principle is similar for <code>on_disc_matrices</code>: single brackets return a subsetted <code>on_disc_matrix</code>, and double brackets return a sparse in-memory matrix.</p>
</div>
<div id="operate-on-an-entire-on_disc_matrix" class="section level2">
<h2 class="hasAnchor">
<a href="#operate-on-an-entire-on_disc_matrix" class="anchor"></a>Operate on an entire <code>on_disc_matrix</code>
</h2>
<p>A common step in many single-cell analysis pipelines is to compute cell-wise and gene-wise summary statistics. The function <code>summarize_expression_matrix</code> takes an <code>on_disc_matrix</code> as an argument and returns some useful cell-wise summary statistics (namely, total UMI count per cell, total number of genes expressed per cell, and percentage of UMIs that are mitochondrial per cell) and gene-wise summary statistics (namely, total UMI count per gene and total number of cells in which gene is expressed). The function works by partitioning the expression data into chunks, loading each chunk into memory, computing the relevant summary statistics for each chunk, and combining the results across chunks. <code>summarize_expression_matrix</code> takes about thirty seconds to run on the real data and about a second to run on the in-package example data. The running time of <code>summarize_expression_matrix</code> is longer than that of other functions because <code>summarize_expression_matrix</code> must load the entire dataset into memory (piece-by-piece).</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">covariate_matrices</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/summarize_expression_matrix.html">summarize_expression_matrix</a></span><span class="op">(</span><span class="va">expression_matrix</span><span class="op">)</span>
<span class="va">cell_covariate_matrix</span> <span class="op">&lt;-</span> <span class="va">covariate_matrices</span><span class="op">$</span><span class="va">cell_covariate_matrix</span>
<span class="va">gene_covariate_matrix</span> <span class="op">&lt;-</span> <span class="va">covariate_matrices</span><span class="op">$</span><span class="va">gene_covariate_matrix</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">cell_covariate_matrix</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">gene_covariate_matrix</span><span class="op">)</span></code></pre></div>
<p>We can use <code>apply</code> to apply more general functions to the rows and columns on an <code>on_disc_matrix</code>. Below, we use <code>apply</code> to calculate the standard deviation and mean of each gene’s expression.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gene_means_and_sds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span>X <span class="op">=</span> <span class="va">expression_matrix</span>, MARGIN <span class="op">=</span> <span class="fl">1</span>, FUN <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">r</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>mean <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">r</span><span class="op">)</span>, sd <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/sd.html">sd</a></span><span class="op">(</span><span class="va">r</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><code>summarize_expression_matrix</code> is much faster than the equivalent calls to <code>apply</code>. Thus, to obtain basic cell-wise and gene-wise summary statistics, it is best to use <code>summarize_expression_matrix</code>.</p>
</div>
<div id="saving-and-loading-on_disc_matrices" class="section level2">
<h2 class="hasAnchor">
<a href="#saving-and-loading-on_disc_matrices" class="anchor"></a>Saving and loading <code>on_disc_matrices</code>
</h2>
<p>It is important to remember that there are two components to an <code>on_disc_matrix</code>: the <code>.h5</code> file stored on disk, and the <code>on_disc_matrix</code> R object stored in memory. The latter contains a file path to the former, allowing users to interact with the on-disk data from within R.</p>
<p>To save an <code>on_disc_matrix</code>, we simply use <code>saveRDS</code> to write the <code>on_disc_matrix</code> R object to an <code>.rds</code> file.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">saveRDS</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">expression_matrix</span>, file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">data_sub_dir</span>, <span class="st">"/expression_matrix.rds"</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/rm.html">rm</a></span><span class="op">(</span><span class="va">expression_matrix</span><span class="op">)</span></code></pre></div>
<p>We can then load the <code>on_disc_matrix</code> by calling <code>readRDS</code> on the <code>.rds</code> file.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">expression_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span>file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">data_sub_dir</span>, <span class="st">"/expression_matrix.rds"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Suppose we forget to save the <code>on_disc_matrix</code> R object that we initialized through a call to <code>create_on_disc_matrix_from_10x_mtx</code>. We can use the function <code>on_disc_matrix</code> (i.e., the constructor of the <code>on_disc_matrix</code> class) to create a new <code>on_disc_matrix</code> by specifying the file path of the corresponding on-disk <code>.h5</code> file.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/rm.html">rm</a></span><span class="op">(</span><span class="va">expression_matrix</span><span class="op">)</span>
<span class="va">expression_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/on_disc_matrix-class.html">on_disc_matrix</a></span><span class="op">(</span>h5_file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">data_sub_dir</span>, <span class="st">"/on_disc_matrix_1.h5"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>You should not call <code>create_on_disc_matrix_from_10x_mtx</code> more than once on the same raw expression data; doing so is wasteful.</p>
</div>
<div id="connecting-to-seurat" class="section level2">
<h2 class="hasAnchor">
<a href="#connecting-to-seurat" class="anchor"></a>Connecting to <code>Seurat</code>
</h2>
<p><code>Seurat</code> is a popular R package for analyzing single-cell data in memory. We can convert an <code>on_disc_matrix</code> object to a <code>Seurat</code> object using the function <code>convert_to_seurat_object</code>. Because the data contained in an <code>on_disc_matrix</code> might not fit entirely into memory, it is often a good idea to first subset the <code>on_disc_matrix</code> before calling <code>convert_to_seurat_object</code>. Below, we define a new <code>on_disc_matrix</code> <code>x</code> that is a subset of <code>expression_matrix</code>. Then, we call <code>convert_to_seurat_object</code> on <code>x</code>.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">x</span> <span class="op">&lt;-</span> <span class="va">expression_matrix</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">100</span>, <span class="op">-</span><span class="op">(</span><span class="fl">300</span><span class="op">:</span><span class="fl">399</span><span class="op">)</span><span class="op">]</span>
<span class="va">seurat_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/convert_to_seurat_object.html">convert_to_seurat_object</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span><span class="op">)</span>
<span class="va">seurat_obj</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://timothy-barry.github.io/">Tim Barry</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
